File: serialization/typescript/text/src/textParseNodeFactory.ts
import { ParseNode, ParseNodeFactory } from "@microsoft/kiota-abstractions";
import { TextDecoder } from "util";

import { TextParseNode } from "./textParseNode";

export class TextParseNodeFactory implements ParseNodeFactory {
  public getValidContentType(): string {
    return "text/plain";
  }
  public getRootParseNode(
    contentType: string,
    content: ArrayBuffer
  ): ParseNode {
    if (!contentType) {
      throw new Error("content type cannot be undefined or empty");
    } else if (this.getValidContentType() !== contentType) {
      throw new Error(`expected a ${this.getValidContentType()} content type`);
    }
    if (!content) {
      throw new Error("content cannot be undefined of empty");
    }
    const decoder = new TextDecoder();
    const contentAsStr = decoder.decode(content);
    return new TextParseNode(contentAsStr);
  }
}

File: serialization/dotnet/text/src/TextParseNodeFactory.cs
using System;
using System.IO;
using Microsoft.Kiota.Abstractions.Serialization;

namespace Microsoft.Kiota.Serialization.Text;

/// <summary>
/// The <see cref="IParseNodeFactory"/> implementation for text/plain content types
/// </summary>
public class TextParseNodeFactory : IParseNodeFactory
{
    /// <inheritdoc />
    public string ValidContentType => "text/plain";
    /// <inheritdoc />
    public IParseNode GetRootParseNode(string contentType, Stream content) {
        if(string.IsNullOrEmpty(contentType))
            throw new ArgumentNullException(nameof(contentType));
        else if(!ValidContentType.Equals(contentType, StringComparison.OrdinalIgnoreCase))
            throw new ArgumentOutOfRangeException($"expected a {ValidContentType} content type");
        
        _ = content ?? throw new ArgumentNullException(nameof(content));
        using var reader = new StreamReader(content);
        var stringContent = reader.ReadToEnd();
        return new TextParseNode(stringContent);
    }
}
